### PR TITLE
Allow underscore-prefixed renamed dependency in manifest

### PR DESCRIPTION
**Pull request by hi-rustin**
close https://github.com/rust-lang/crates.io/issues/3532

### CODE REVIEW

**Comment by Turbo87** [comment]
👋 @hi-rustin 

the issue here is slightly more complicated. crate name generally are not allowed to start with an underscore, so adjusting `valid_ident()` to allow it could create some problems for us when people suddenly start publishing crates with a leading underscore.

https://github.com/rust-lang/crates.io/pull/1487 implemented support for "renamed dependencies", but it looks like we did not test for the "`name` has a leading underscore" case. I think the first step in fixing this issue is to create a new test, roughly similar to the existing `new_with_renamed_dependency()` test, but with a leading underscore for the `name`. That should hopefully reproduce the issue that @jmagnuson is seeing in #3532.

**Comment by hi-rustin** [comment]
> I think the first step in fixing this issue is to create a new test, roughly similar to the existing `new_with_renamed_dependency()` test, but with a leading underscore for the `name`.

Added.

**Comment by hi-rustin** [comment]
The test case will fail when there is no such underlined change.

```log
failed to decode: Error("missing field `crate`", line: 1, column: 249)
thread 'krate::publish::new_with_underscore_name_dependency' panicked at 'failed to decode: Error("missing field `crate`", line: 1, column: 249)', src/tests/util/response.rs:111:19
stack backtrace:
   0: rust_begin_unwind
             at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/std/src/panicking.rs:493:5
   1: std::panicking::begin_panic_fmt
             at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/std/src/panicking.rs:435:5
   2: all::util::response::json
             at ./src/tests/util/response.rs:111:19
   3: all::util::response::Response<T>::good
             at ./src/tests/util/response.rs:26:9
   4: all::krate::publish::new_with_underscore_name_dependency
             at ./src/tests/krate/publish.rs:187:5
   5: all::krate::publish::new_with_underscore_name_dependency::{{closure}}
             at ./src/tests/krate/publish.rs:174:1
   6: core::ops::function::FnOnce::call_once
             at /Users/rustin/.rustup/toolchains/stable-x86_64-apple-darwin/lib/rustlib/src/rust/library/core/src/ops/function.rs:227:5
   7: core::ops::function::FnOnce::call_once
             at /rustc/9bc8c42bb2f19e745a63f3445f1ac248fb015e53/library/core/src/ops/function.rs:227:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: test failed, to rerun pass '-p cargo-registry --test all'

Process finished with exit code 101
```

**Comment by hi-rustin** [comment]
I did some local debugging and integration testing to make sure the logic was being tested.

File: src/tests/krate/publish.rs
Comment (Turbo87) [inline src/tests/krate/publish.rs:247]: `_foo-dep` is not a legal crate name, so this call should actually fail, because in this case we were not renaming the dependency

Comment (hi-rustin) [reply src/tests/krate/publish.rs:247]: Oh, I see what you mean! _ can only be used in renaming.

Comment (Turbo87) [reply src/tests/krate/publish.rs:247]: yeah, exactly

Comment (Turbo87) [reply src/tests/krate/publish.rs:247]: I'm not 100% sure, but I think `explicit_name_in_toml: Option<EncodableCrateName>` is the issue, and maybe it should be `explicit_name_in_toml: Option<String>` instead, or at least some type that does allow underscore prefixes.

Comment (hi-rustin) [reply src/tests/krate/publish.rs:247]: Yes, you are right! I have added it. I'll sort out the code and submit it tomorrow!

**Review by Turbo87 — APPROVED** [review]
looks good to me now, but I would like to get a second pair of eyes on this before we merge it.

@pietroalbini @jtgeibel @JohnTitor

**Review by pietroalbini — CHANGES_REQUESTED** [review]
The code changes themselves look great, but I'd prefer for the internal API to be harder to misuse and misunderstand:

* The boolean `Crate::valid_name("foo", true)` or `Crate::valid_name("bar" false)` is a bit confusing if one doesn't know about this PR. I'd prefer to keep `Crate::valid_name` the same as before and adding a `Crate::valid_dependency_name`.
* `EncodableExplicitName` is also a bit confusing, it'd be better to call it `EncodableDependencyName`.

**Comment by hi-rustin** [comment]
> The code changes themselves look great, but I'd prefer for the internal API to be harder to misuse and misunderstand:
> 
>     * The boolean `Crate::valid_name("foo", true)` or `Crate::valid_name("bar" false)` is a bit confusing if one doesn't know about this PR. I'd prefer to keep `Crate::valid_name` the same as before and adding a `Crate::valid_dependency_name`.
> 
>     * `EncodableExplicitName` is also a bit confusing, it'd be better to call it `EncodableDependencyName`.

Addressed! Thanks for your review.

**Review by pietroalbini — APPROVED** [review]
Thanks!

**Comment by jmagnuson** [comment]
Thank you, @hi-rustin, and everyone who helped get this merged!

**Pull request merged**
